river/LayoutManager.zig

```diff
@@ -0,0 +1,378 @@
+pub const LayoutDemand = struct {
+    serial: u32 = 0,
+    views: u32 = 0,
+    output: *Output = undefined,
+
+    /// Timeout timer
+    timer: ?*wl.EventSource = null,
+
+    /// Proposed view dimensions
+    view_dimensions: std.TailQueue(Box) = .{},
+
+    /// Arbitrary timeout
+    const timeout: u32 = 1000;
+
+    pub fn init(self: *LayoutDemand, output: *Output, views: u32) void {
+        self.* = .{
+            .serial = output.root.server.wl_server.nextSerial(),
+            .output = output,
+            .views = views,
+        };
+
+        // Attach a timout timer
+        const event_loop = output.root.server.wl_server.getEventLoop();
+        self.timer = event_loop.addTimer(*LayoutDemand, handleTimeout, self) catch {
+            log.err(.view, "failed to create layout timeout timer", .{});
+            return;
+        };
+        self.timer.?.timerUpdate(timeout) catch |err| {
+            log.err(.layout_manager, "failed to update layout timeout timer: {}", .{err});
+            self.timer.?.remove();
+            self.timer = null;
+        };
+    }
+
+    pub fn deinit(self: *LayoutDemand) void {
+        if (self.timer) |timer| timer.remove();
+
+        while (self.view_dimensions.pop()) |node| {
+            util.gpa.destroy(node);
+        }
+    }
+
+    /// Destroy the LayoutDemand on timeout.
+    /// All further responses to the event will simply be ignored.
+    fn handleTimeout(self: *LayoutDemand) callconv(.C) c_int {
+        log.notice(
+            .layout_manager,
+            "layout demand for output '{}' timed out",
+            .{self.output.wlr_output.name},
+        );
+        self.deinit();
+        const output = self.output;
+        if (output.layout_demand) |demand| {
+            util.gpa.destroy(demand);
+            output.layout_demand = null;
+        }
+
+        // Start a transaction, so things that happened while
+        // this LayoutDemand was timing out are applied correctly.
+        // Remember: An active layout demand blocks all transactions.
+        // on the affected output.
+        output.root.startTransaction();
+
+        return 0;
+    }
+
+    /// Push a set of proposed view dimensions and position to the list
+    pub fn pushViewDimensions(self: *LayoutDemand, x: i32, y: i32, width: u32, height: u32) !void {
+        const node = try util.gpa.create(std.TailQueue(Box).Node);
+
+        // Here we apply the offset to align the coords with the origin of the
+        // usable area and shrink the dimensions to accomodate the border size.
+        node.data = .{
+            .x = x + self.output.usable_box.x + @intCast(i32, self.output.root.server.config.border_width),
+            .y = y + self.output.usable_box.y + @intCast(i32, self.output.root.server.config.border_width),
+            .width = width - 2 * self.output.root.server.config.border_width,
+            .height = height - 2 * self.output.root.server.config.border_width,
```

Comment:
This will underflow if width is small and border width is high. We need to handle this case.